src/Workspaces/MSBuildTest/VisualStudioMSBuildWorkspaceTests.cs

```diff
@@ -3251,6 +3251,44 @@ public async Task MSBuildProjectShouldHandleCodePageProperty()
             Assert.Equal("//\u00E2\u20AC\u0153".Length, text.Length);
         }
 
+        [ConditionalFact(typeof(VisualStudioMSBuildInstalled))]
+        public async Task MSBuildWorkspaceDocumentsFoldersProperty()
+        {
+            CreateFiles(GetNetCoreAppFiles()
+                .WithFile(@"dir1\dir2\dir3\MyClass.cs", Resources.SourceFiles.CSharp.CSharpClass));
+
+            using var workspace = CreateMSBuildWorkspace();
+            var project = await workspace.OpenProjectAsync(GetSolutionFileName("Project.csproj"));
+            var document = project.Documents.Single(d => d.Name == "MyClass.cs");
+            Assert.Equal(new[] { "dir1", "dir2", "dir3" }, document.Folders);
+        }
+
+        [ConditionalFact(typeof(VisualStudioMSBuildInstalled))]
+        public async Task MSBuildWorkspaceLinkedDocumentHasFolders()
+        {
+            CreateFiles(GetSimpleCSharpSolutionFiles()
+                .WithFile(@"CSharpProject\CSharpProject.csproj", Resources.ProjectFiles.CSharp.WithLink)
+                .WithFile(@"OtherStuff\Foo.cs", Resources.SourceFiles.CSharp.OtherStuff_Foo));
+
+            using var workspace = CreateMSBuildWorkspace();
+            var project = await workspace.OpenProjectAsync(GetSolutionFileName(@"CSharpProject\CSharpProject.csproj"));
+            var linkedDocument = project.Documents.Single(d => d.Name == "Foo.cs");
+            Assert.Equal(new[] { "Blah" }, linkedDocument.Folders);
+        }
+
+        [ConditionalFact(typeof(VisualStudioMSBuildInstalled))]
+        public async Task MSBuildWorkspaceWithDocumentInParentFolders()
+        {
+            CreateFiles(GetSimpleCSharpSolutionFiles()
+                .WithFile(@"CSharpProject\CSharpProject.csproj", Resources.ProjectFiles.CSharp.WithClassNotInProjectFolder)
+                .WithFile(@"MyDir\MyClass.cs", Resources.SourceFiles.CSharp.CSharpClass));
+
+            using var workspace = CreateMSBuildWorkspace();
+            var project = await workspace.OpenProjectAsync(GetSolutionFileName(@"CSharpProject\CSharpProject.csproj"));
+            var linkedDocument = project.Documents.Single(d => d.Name == "MyClass.cs");
+            Assert.Equal(new[] { "..", "MyDir" }, linkedDocument.Folders);
```

Comment:
If you try looking at the workspace we get in Visual Studio, does it return .. entries like this?

Reply:
I think yes. Tried to debug from the code action provider.
![image](https://github.com/dotnet/roslyn/assets/24360909/3000f262-168a-4fe0-8794-557b439fc304)